### **Soal 1 Pandas Data Frame**

In [1]:
from sklearn.datasets import make_regression
# generate dataset
X, y = make_regression(n_samples=100, n_features=100, n_informative=10)

In [2]:
lists= ["X{0}".format(i) for i in range (0,101)]  
#list=''.join(lists).split(',')
lists[100]='Y'
lists

['X0',
 'X1',
 'X2',
 'X3',
 'X4',
 'X5',
 'X6',
 'X7',
 'X8',
 'X9',
 'X10',
 'X11',
 'X12',
 'X13',
 'X14',
 'X15',
 'X16',
 'X17',
 'X18',
 'X19',
 'X20',
 'X21',
 'X22',
 'X23',
 'X24',
 'X25',
 'X26',
 'X27',
 'X28',
 'X29',
 'X30',
 'X31',
 'X32',
 'X33',
 'X34',
 'X35',
 'X36',
 'X37',
 'X38',
 'X39',
 'X40',
 'X41',
 'X42',
 'X43',
 'X44',
 'X45',
 'X46',
 'X47',
 'X48',
 'X49',
 'X50',
 'X51',
 'X52',
 'X53',
 'X54',
 'X55',
 'X56',
 'X57',
 'X58',
 'X59',
 'X60',
 'X61',
 'X62',
 'X63',
 'X64',
 'X65',
 'X66',
 'X67',
 'X68',
 'X69',
 'X70',
 'X71',
 'X72',
 'X73',
 'X74',
 'X75',
 'X76',
 'X77',
 'X78',
 'X79',
 'X80',
 'X81',
 'X82',
 'X83',
 'X84',
 'X85',
 'X86',
 'X87',
 'X88',
 'X89',
 'X90',
 'X91',
 'X92',
 'X93',
 'X94',
 'X95',
 'X96',
 'X97',
 'X98',
 'X99',
 'Y']

In [3]:
import pandas as pd
import numpy as np
df = pd.DataFrame(np.random.rand(100, 101), columns=lists)
df


X0        X1        X2        X3        X4        X5        X6  \
0   0.304079  0.757581  0.826305  0.639717  0.479173  0.398221  0.233098   
1   0.859872  0.247374  0.874206  0.345943  0.864720  0.404596  0.010799   
2   0.379072  0.098238  0.251553  0.135129  0.049096  0.346759  0.678579   
3   0.792472  0.134590  0.116994  0.629079  0.457152  0.050090  0.156244   
4   0.783647  0.300961  0.174594  0.610759  0.365026  0.359666  0.958065   
..       ...       ...       ...       ...       ...       ...       ...   
95  0.351828  0.654395  0.935986  0.080469  0.367924  0.762201  0.438042   
96  0.294818  0.318703  0.906913  0.743923  0.090336  0.129194  0.085781   
97  0.097973  0.259377  0.853230  0.737573  0.885796  0.858281  0.457617   
98  0.543861  0.683335  0.319561  0.171260  0.498527  0.273059  0.038690   
99  0.037123  0.642737  0.806018  0.538620  0.494495  0.134385  0.327315   

          X7        X8        X9  ...       X91       X92       X93       X94  \
0   0.102799  0.669075  0.654486  ...  0.819733  0.342934  0.201649  0.445187   
1   0.718618  0.734606  0.430939  ...  0.574497  0.960069  0.084286  0.001594   
2   0.133999  0.118342  0.295459  ...  0.286881  0.205852  0.780054  0.203339   
3   0.888372  0.394259  0.632841  ...  0.062709  0.319598  0.725906  0.439296   
4   0.289691  0.026666  0.927150  ...  0.864159  0.263219  0.027035  0.430277   
..       ...       ...       ...  ...       ...       ...       ...       ...   
95  0.989799  0.379282  0.708902  ...  0.541637  0.877375  0.117976  0.946766   
96  0.445283  0.925584  0.176880  ...  0.631917  0.923734  0.373532  0.500906   
97  0.920799  0.693205  0.362934  ...  0.747307  0.713408  0.967755  0.601706   
98  0.260784  0.396264  0.102316  ...  0.369245  0.824899  0.025566  0.313539   
99  0.462916  0.043664  0.745797  ...  0.615470  0.454031  0.333034  0.233424   

         X95       X96       X97       X98       X99         Y  
0   0.695252  0.012177  0.470954  0.916246  0.790817  0.920340  
1   0.231055  0.083560  0.997041  0.178282  0.855410  0.991730  
2   0.351921  0.405452  0.650321  0.171997  0.062778  0.336984  
3   0.013522  0.290641  0.083744  0.079000  0.841830  0.988492  
4   0.871651  0.559073  0.741637  0.605433  0.294561  0.231899  
..       ...       ...       ...       ...       ...       ...  
95  0.457081  0.090818  0.198168  0.416645  0.148359  0.099885  
96  0.332880  0.099013  0.379326  0.216701  0.391088  0.644806  
97  0.565803  0.160790  0.005737  0.876544  0.155968  0.578465  
98  0.382737  0.064731  0.513890  0.226230  0.811175  0.969502  
99  0.217107  0.727819  0.177320  0.959396  0.482427  0.403427  

[100 rows x 101 columns]

Buatlah sebuah Dataframe dengan data di atas. Dimana terdapat variabel X dan y. buatlah menjadi column X0-X99 dan y pada akhir column.

In [4]:
#jangan di run
#cell contoh hasil

### Soal 02

Berapa banyak perenang yang memiliki medali perunggu(bronze)?

In [5]:
import pandas as pd
df=pd.read_csv('summer.csv')
df

Year    City      Sport           Discipline  \
0      1896  Athens   Aquatics             Swimming   
1      1896  Athens   Aquatics             Swimming   
2      1896  Athens   Aquatics             Swimming   
3      1896  Athens   Aquatics             Swimming   
4      1896  Athens   Aquatics             Swimming   
...     ...     ...        ...                  ...   
31160  2012  London  Wrestling  Wrestling Freestyle   
31161  2012  London  Wrestling  Wrestling Freestyle   
31162  2012  London  Wrestling  Wrestling Freestyle   
31163  2012  London  Wrestling  Wrestling Freestyle   
31164  2012  London  Wrestling  Wrestling Freestyle   

                         Athlete Country Gender                       Event  \
0                  HAJOS, Alfred     HUN    Men              100M Freestyle   
1               HERSCHMANN, Otto     AUT    Men              100M Freestyle   
2              DRIVAS, Dimitrios     GRE    Men  100M Freestyle For Sailors   
3             MALOKINIS, Ioannis     GRE    Men  100M Freestyle For Sailors   
4             CHASAPIS, Spiridon     GRE    Men  100M Freestyle For Sailors   
...                          ...     ...    ...                         ...   
31160         JANIKOWSKI, Damian     POL    Men                    Wg 84 KG   
31161  REZAEI, Ghasem Gholamreza     IRI    Men                    Wg 96 KG   
31162             TOTROV, Rustam     RUS    Men                    Wg 96 KG   
31163          ALEKSANYAN, Artur     ARM    Men                    Wg 96 KG   
31164             LIDBERG, Jimmy     SWE    Men                    Wg 96 KG   

        Medal  
0        Gold  
1      Silver  
2      Bronze  
3        Gold  
4      Silver  
...       ...  
31160  Bronze  
31161    Gold  
31162  Silver  
31163  Bronze  
31164  Bronze  

[31165 rows x 9 columns]

In [6]:
#your code goes here
Medal=df[df['Medal']=='Bronze']
len(Medal)

10369

Answer: 10369


Dalam olahraga air (Aquatics), kota mana yang memiliki rata-rata perolehan medali tertinggi?

In [7]:
#your code goes here
df[df['Sport']=='Aquatics']['City'].max()

'Tokyo'

Answer: Tokyo

### **Soal 3 List Comprehension & Pandas** 

Pada Soal ini kalian **wajib** menggunakan feature list comprehension dan juga apply.<br>

In [8]:
import pandas as pd
df=pd.read_csv('Salaries.csv')
df

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651       Not provided   
148651  148652       Not provided   
148652  148653       Not provided   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                    Not provided        NaN   
148651                                    Not provided        NaN   
148652                                    Not provided        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25       NaN  567595.43         567595.43  2011   
1         245131.88  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649         0.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653         0.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

PT.A baru saja kena tegur dari negara karena ketauan tidak membayar uang lemburan. Anda sebagai seorang data Scientist diminta PT.A untuk mengolah data yang ada.

**Step pertama**

**Wajib Menggunakan .Apply()** <br>
PT.A melihat beberapa data tidak di isi oleh sebagian karyawan. kalian di minta untuk untuk merubah nilai 'Not provided' pada column `JobTitle` dan `EmployeeName` menjadi 'Error'.

In [17]:
df[['JobTitle', 'EmployeeName']]=df[['JobTitle', 'EmployeeName']].apply(lambda x:x.replace({'Not provided': 'Eror'}))
df

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651               Eror   
148651  148652               Eror   
148652  148653               Eror   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                            Eror        NaN   
148651                                            Eror        NaN   
148652                                            Eror        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25       NaN  567595.43         567595.43  2011   
1         245131.88  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649         0.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653         0.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


In [ ]:
#jangan tulis code disini
#cell contoh hasil:

**Step Kedua**

**WAJIB Gunakan list comprehension**<br>
Anda diminta untuk menambahkan 200000 pada karyawan yang mendapat Overtime 0 pada Data yang telah dibersihkan di tahap 1.


In [55]:
df['OvertimePay']=df[(df['OvertimePay'].astype(str))]
df[(df['OvertimePay'].apply(lambda x: x.replace({'0.00', '200000'})))]
df

KeyError: "None of [Index(['0.0', '245131.88', '106088.18', '56120.71', '9737.0', '8601.0',\n       '89062.9', '0.0', '86362.68', '0.0',\n       ...\n       '0.0', '0.0', 'nan', '0.0', '0.0', '0.0', 'nan', 'nan', 'nan', '0.0'],\n      dtype='object', length=148654)] are in the [columns]"

In [ ]:
#jangan tulis code disini
#cell contoh hasil:

**Tahap ketiga** 

**bebas memakai teknik apapun**<br>
selanjutnya ada kelasahan pada penulisan 'CAPTAIN III (POLICE DEPARTMENT)' ubahlah menjadi 'CAPTAIN III' saja

In [20]:
#your code goes here
df['JobTitle']= df['JobTitle'].replace('CAPTAIN III (POLICE DEPARTMENT)', 'CAPTAIN III')
df

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651               Eror   
148651  148652               Eror   
148652  148653               Eror   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                                          CAPTAIN III  155966.02   
2                                          CAPTAIN III  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                            Eror        NaN   
148651                                            Eror        NaN   
148652                                            Eror        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25       NaN  567595.43         567595.43  2011   
1         245131.88  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649         0.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653         0.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

In [ ]:
#jangan tulis code disini
#cell contoh hasil:

### **Soal 4 One Hot Encoding**

Buatlah **sebuah fungsi** bernama `one_hot`, yang menerima input sebuah numpy array.

Ex:<br>
input array [2,3,1,2] maka hasil dari fungsi kalian:

array [[0,1,0], [0,0,1], [1,0,0], [0,1,0]]

Constrain: <br>

- Integers dalam input selalu di awali 1. 
- Integers dalam input hanya positif dan bukan 0.
- Integers dalam input tidak akan skip angka.

In [60]:
import numpy as np
input_array = np.array([3,2,1,2,3,3,3])
#your code goes here
def one_hot(input_array):
            b = np.zeros((input_array.size, input_array.max()), dtype=int)
            b[np.arange(input_array.size), (input_array)-1] =1
            return b
input_array = np.array([3,2,1,2,3,3,3])
one_hot(input_array)


    


array([[0, 0, 1],
       [0, 1, 0],
       [1, 0, 0],
       [0, 1, 0],
       [0, 0, 1],
       [0, 0, 1],
       [0, 0, 1]])

In [ ]:
#jangan tulis code disini
#cell contoh hasil: